### ISSUE TITLE
About torch version

### ISSUE THREAD

**Issue by chooron** [ckrapu/gr4j_theano#1]
Hello, about your work in "**Gradient‐Based Inverse Estimation for a Rainfall‐Runoff Model**“
I am very interested in the approach presented in the article and I am grateful for the work you have done.
I converted the theano version to the pytorch version according to the code you provided, while referring to the gr4j model in https://github.com/amacd31/gr4j. Bringing your parameters into my pytorch code gives the same result. However, probably due to some differences between torch and theano, and some computational steps of the gr4j model that are not suitable for automatic derivation, my code cannot implement algorithms such as automatic derivation, gradient descent, etc.
To this end, if you also have a partial understanding of torch, I hope you can point me to the problem of the code

**Comment by ckrapu** [comment]
Hi chooron, I actually did recently implement GR4J in PyTorch for use in NumPyro later. You can find that here: https://github.com/ckrapu/tvp-gr4j/blob/main/gr4j_jax.py. Let me know if you have questions about it.

**Comment by chooron** [comment]
Oh thank you, I'll look at this code later

**Comment by chooron** [comment]
Hi, ckrapu, I have solved this problem and converted your code into torch version. It can train like the network. I am deeply inspired by your work. Now, I want to know if part of GR4J can be replaced by torch.nn.layers, which may improve prediction performance.

**Comment by ckrapu** [comment]
Without knowing too much more about that, I would probably guess that it would be possible to replace some of the simpler functions with layers. You may want to look at a slightly more complex model as well since GR4J replaces parameterized functions with hard-coded step functions or cutoffs in some places.
